# Numpy Ops

In [11]:
import numpy as np

## [dot product](https://numpy.org/doc/stable/reference/generated/numpy.dot.html) ([kr](https://runebook.dev/ko/docs/numpy/reference/generated/numpy.dot))
```
dot(a, b)[i,j,k,m] = sum(a[i,j,:] * b[k,:,m])
```
* 첫 행렬의 마지막 차원과, 두번째 행렬의 끝에서 두번째 차원의 크기가 동일 해야함
* 연산 결과 두 행렬의 연산 차원이 없어지며, 두 행렬의 차원이 합쳐짐
    - eg> $inner(x(5,4,3,2), y(6,5,4,2,3) = z(5,4,3,6,5,4,3)$

cf> matmul product

---
### $dot(x(k,), y(k,)) = z()$
$\begin{align}
np.dot(x, y) = np.dot(
    \begin{bmatrix} 1 & 2 \end{bmatrix}
    ,
    \begin{bmatrix} 10 \\ 20 \end{bmatrix}
)
&= 1 \times 10 + 2 \times 20 \\
&= 50
\end{align}$

In [12]:
x = np.array([1, 2])
print(f"x{x.shape} =", x, "\n")

y = np.array([10, 20])
print(f"y{y.shape} =", y, "\n")

z = np.dot(x, y)
print(f"np.dot(x, y) = z{z.shape} =", z, "\n")

z = np.matmul(x, y)
print(f"np.matmul(x, y) = z{z.shape} =", z, "\n")

x(2,) = [1 2] 

y(2,) = [10 20] 

np.dot(x, y) = z() = 50 

np.matmul(x, y) = z() = 50 



---
### $dot(x(l, k), y(k, n)) = z(l, n)$
$\begin{align}
np.dot(x, y) = np.dot(
    \begin{bmatrix}
        1 & 2 \\
        3 & 4 \\
        5 & 6 \\
    \end{bmatrix}
    ,
    \begin{bmatrix}
        10 & 20 & 30 \\
        40 & 50 & 60 \\
    \end{bmatrix}
)
&=  \begin{bmatrix}
        1 \times 10 + 2 \times 40 & 1 \times 20 + 2 \times 50 & 1 \times 30 + 2 \times 60 \\
        3 \times 10 + 4 \times 40 & 3 \times 20 + 4 \times 50 & 3 \times 30 + 4 \times 60 \\
        5 \times 10 + 6 \times 40 & 5 \times 20 + 6 \times 50 & 5 \times 30 + 6 \times 60 \\
    \end{bmatrix} \\
&=  \begin{bmatrix}
         90 & 120 & 150 \\
        190 & 360 & 330 \\
        290 & 400 & 510 \\
    \end{bmatrix}
\end{align}$

In [13]:
x = np.array([
    [1, 2],
    [3, 4],
    [5, 6]
])
print(f"x{x.shape} =", x, "\n")

y = np.array([
    [10, 20, 30],
    [40, 50, 60],
])
print(f"y{y.shape} =", y, "\n")

z = np.dot(x, y)
print(f"np.dot(x, y) = z{z.shape} =", z, "\n")

z = np.matmul(x, y)
print(f"np.matmul(x, y) = z{z.shape} =", z, "\n")

x(3, 2) = [[1 2]
 [3 4]
 [5 6]] 

y(2, 3) = [[10 20 30]
 [40 50 60]] 

np.dot(x, y) = z(3, 3) = [[ 90 120 150]
 [190 260 330]
 [290 400 510]] 

np.matmul(x, y) = z(3, 3) = [[ 90 120 150]
 [190 260 330]
 [290 400 510]] 



---
### $dot(x(l, ..., k), y(n, ..., k, m)) = z(l,..., n, ..., m)$
$\begin{align}
np.dot(x, y) = np.dot(
\begin{bmatrix}
    \begin{bmatrix} 1 & 2 \end{bmatrix} \\
    \begin{bmatrix} 3 & 4 \end{bmatrix} \\
    \begin{bmatrix} 5 & 6 \end{bmatrix}
\end{bmatrix}
,
\begin{bmatrix}
    \begin{bmatrix}
        \begin{bmatrix} 10 & 20 & 30 \end{bmatrix} \\
        \begin{bmatrix} 40 & 50 & 60 \end{bmatrix}
    \end{bmatrix} \\
    \begin{bmatrix}
        \begin{bmatrix} 100 & 200 & 300 \end{bmatrix} \\
        \begin{bmatrix} 400 & 500 & 600 \end{bmatrix}
    \end{bmatrix}
\end{bmatrix}
)
&= \begin{bmatrix}
    \begin{bmatrix}
        \begin{bmatrix}
            1 \times 10 + 2 \times 40 & 1 \times 20 + 2 \times 50 & 1 \times 30 + 2 \times 60
        \end{bmatrix} \\
        \begin{bmatrix}
            1 \times 100 + 2 \times 400 & 1 \times 200 + 2 \times 500 & 1 \times 300 + 2 \times 600
        \end{bmatrix}
    \end{bmatrix} \\
    \begin{bmatrix}
        \begin{bmatrix}
            3 \times 10 + 4 \times 40 & 3 \times 20 + 4 \times 50 & 3 \times 30 + 4 \times 60
        \end{bmatrix} \\
        \begin{bmatrix}
            3 \times 100 + 4 \times 400 & 3 \times 200 + 4 \times 500 & 3 \times 300 + 4 \times 600
        \end{bmatrix}
    \end{bmatrix} \\
    \begin{bmatrix}
        \begin{bmatrix}
            5 \times 10 + 6 \times 40 & 5 \times 20 + 6 \times 50 & 5 \times 30 + 6 \times 60
        \end{bmatrix} \\
        \begin{bmatrix}
            5 \times 100 + 6 \times 400 & 5 \times 200 + 6 \times 500 & 5 \times 300 + 6 \times 600
        \end{bmatrix}
    \end{bmatrix}
\end{bmatrix} \\
&= \begin{bmatrix}
    \begin{bmatrix}
        \begin{bmatrix}
            90 & 120 & 150
        \end{bmatrix} \\
        \begin{bmatrix}
            900 & 1200 & 1500
        \end{bmatrix}
    \end{bmatrix} \\
    \begin{bmatrix}
        \begin{bmatrix}
            190 & 260 & 330
        \end{bmatrix} \\
        \begin{bmatrix}
            1900 & 2600 & 3300
        \end{bmatrix}
    \end{bmatrix} \\
    \begin{bmatrix}
        \begin{bmatrix}
            290 & 400 & 510
        \end{bmatrix} \\
        \begin{bmatrix}
            2900 & 4000 & 5100
        \end{bmatrix}
    \end{bmatrix}
\end{bmatrix}
\end{align}$

In [19]:
x = np.array([
    [1, 2],
    [3, 4],
    [5, 6],
])
print(f"x{x.shape} =", x, "\n")

y = np.array([
    [
        [10, 20, 30],
        [40, 50, 60],
    ],
    [
        [100, 200, 300],
        [400, 500, 600],
    ]
])
print(f"y{y.shape} =", y, "\n")

z = np.dot(x, y)
print(f"np.dot(x, y) = z{z.shape} =", z, "\n")

try:
    z = x * y
    print(f"x * y = z{z.shape} =", z, "\n")
except Exception as e:
    print("Error : x * y = ", e, "\n")

try:
    z = x @ y
    print(f"x @ y = z{z.shape} =", z, "\n")
except Exception as e:
    print("Error : x @ y =", e, "\n")

try:
    z = np.matmul(x, y)
    print(f"np.matmul(x, y) = z{z.shape} =", z, "\n")
except Exception as e:
    print("Error : np.matmul(x, y) =", e, "\n")

x(3, 2) = [[1 2]
 [3 4]
 [5 6]] 

y(2, 2, 3) = [[[ 10  20  30]
  [ 40  50  60]]

 [[100 200 300]
  [400 500 600]]] 

np.dot(x, y) = z(3, 2, 3) = [[[  90  120  150]
  [ 900 1200 1500]]

 [[ 190  260  330]
  [1900 2600 3300]]

 [[ 290  400  510]
  [2900 4000 5100]]] 

Error : x * y =  operands could not be broadcast together with shapes (3,2) (2,2,3)  

x @ y = z(2, 3, 3) = [[[  90  120  150]
  [ 190  260  330]
  [ 290  400  510]]

 [[ 900 1200 1500]
  [1900 2600 3300]
  [2900 4000 5100]]] 

np.matmul(x, y) = z(2, 3, 3) = [[[  90  120  150]
  [ 190  260  330]
  [ 290  400  510]]

 [[ 900 1200 1500]
  [1900 2600 3300]
  [2900 4000 5100]]] 

